File: 2024/day-02/day2.py
def check_report(report):
  is_increasing = int(report[0]) < int(report[1])

  for i in range(1, len(report)):
    i1 = int(report[i - 1])
    i2 = int(report[i])
    diff = abs(i2 - i1)
    is_invalid_change = diff < 1 or diff > 3

    if (is_increasing and i1 > i2) or (not is_increasing and i1 < i2) or is_invalid_change:
      return (True, i - 1)
  return (False, -1)

def solve_part1():
  unsafe_counter = 0
  total_lines = 0
  problematic_reports = []
  problematic_indexes = []

  with open("./day-02/input.txt", "r") as f:
    for line in f:
      total_lines += 1
      report = line.split()
      is_report_problematic, index = check_report(report)
      if (is_report_problematic):
        unsafe_counter += 1
        problematic_reports.append(report)
        problematic_indexes.append(index)
  return problematic_reports, problematic_indexes, total_lines - unsafe_counter

def solve_part2(problematic_reports, problematic_indexes):
  # Safe if there is only 1 problematic level in a report.
  # Naive brute force: is to remove each level in a report and check if the report is still problematic.
  # -> Cost of checking updated report = cost of checking report * O(n) where n is the size of the report = O(n^2)

  # Slightly more optimized: Pass the index of the problematic level. Remove the problematic level or its surrounding levels, 
  # then check if the report is still problematic.
  # -> Cost of checking updated report = cost of checking report * O(1) = O(n)
  fixed_reports = 0
  for report, index in zip(problematic_reports, problematic_indexes):
    report1 = report[:index - 1] + report[index:] # excludes report[index - 1]
    report2 = report[:index] + report[index + 1:] # excludes report[index]
    report3 = report[:index + 1] + report[index + 2:] # excludes report[index + 1]
    
    if not check_report(report1)[0] or not check_report(report2)[0] or not check_report(report3)[0]:
      print(report1, report2, report3)
      fixed_reports += 1

  return part1_solution + fixed_reports

if __name__ == "__main__":
  problematic_reports, problematic_indexes, part1_solution = solve_part1()
  print(part1_solution)
  print(solve_part2(problematic_reports, problematic_indexes))

File: 2024/day-02/day2.go
package main

import (
  "bufio"
  "fmt"
  "os"
  "strings"
  "strconv"
)

func process_input() [][]int {
  var result [][]int

  file, _ := os.Open("./day-02/input.txt")
  defer file.Close()

  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    line := strings.TrimSpace(scanner.Text())
    stringParts := strings.Fields(line)

    var intParts []int = make([]int, len(stringParts))
    for i, str := range stringParts {
      intParts[i], _ = strconv.Atoi(str)
    }
    result = append(result, intParts)
  }
  return result
}

func isSorted(ls []int) (bool, int) {
  if len(ls) <= 1 {
    return true, -1
  }

  var isSortDirectionReversed bool = ls[0] > ls[1]
  for i := 1; i < len(ls); i++ {
    value1 := ls[i - 1]
    value2 := ls[i]
    if isSortDirectionReversed {
      if value1 <= value2 {
        return false, i - 1
      }
    } else {
      if value1 >= value2 {
        return false, i - 1
      }
    }
  }
  
  return true, -1
}

func isDifferenceAcceptable(ls []int) (bool, int) {
  for i := 1; i < len(ls); i++ {
    difference := abs(ls[i] - ls[i - 1])
    if difference < 1 || difference > 3 {
      return false, i
    }
  }
  return true, -1
}

func abs(x int) int {
  if x < 0 {
    return -x
  }
  return x
}

func isReportProblematic(report []int) (bool, int) {
  isSorted, i1 := isSorted(report);
  isDifferenceAcceptable, i2 := isDifferenceAcceptable(report); 

  if !isSorted {
    return true, i1
  }
  if !isDifferenceAcceptable {
    return true, i2
  }
  return false, -1
}

func solvePart1(input [][]int) (int, [][]int, []int) {
  var total int = 0
  var problematicReports [][]int = make([][]int, 0)
  var problematicIndexes []int = make([]int, 0)

  for _, ls := range input {
    isReportProblematic, i := isReportProblematic(ls)
    if isReportProblematic {
      problematicReports = append(problematicReports, ls)
      problematicIndexes = append(problematicIndexes, i)
    } else {
      total += 1
    }
  }
  return total, problematicReports, problematicIndexes
}

func solvePart2(problematicReports [][]int, problematicIndexes []int, part1Solution int) int {
  if len(problematicReports) != len(problematicIndexes) {
    return -1
  }

  total := part1Solution
  for i := 0; i < len(problematicReports); i++ {
    var report []int = problematicReports[i]
    var index int = problematicIndexes[i]

    var fixedReports [][]int
    if index > 0 {
      reportCopy := make([]int, len(report))
      copy(reportCopy, report)
      fixedReport := append(reportCopy[:index - 1], reportCopy[index:]...)
      fixedReports = append(fixedReports, fixedReport)
    }
    if index < len(report) {
      reportCopy := make([]int, len(report))
      copy(reportCopy, report)
      fixedReport := append(reportCopy[:index], reportCopy[index + 1:]...)
      fixedReports = append(fixedReports, fixedReport)
    }
    if index + 1 < len(report) {
      reportCopy := make([]int, len(report))
      copy(reportCopy, report)
      fixedReport := append(reportCopy[:index + 1], reportCopy[index + 2:]...)
      fixedReports = append(fixedReports, fixedReport)
    }

    for _, fixedReport := range fixedReports {
      problematic, _ := isReportProblematic(fixedReport);
      if !problematic {
        total += 1
        break
      }
    }
  }
  return total
}

func main() {
  var input [][]int = process_input()
  part1Solution, problematicReports, problematicIndexes := solvePart1(input)
  fmt.Println("Part 1:", part1Solution)
  fmt.Println("Part 2:", solvePart2(problematicReports, problematicIndexes, part1Solution))
}
